### ISSUE TITLE
Writing to an existing file

### ISSUE THREAD

**Issue by sandstrom** [felixbuenemann/xlsxtream#6]
When writing to an existing file, the file is closed after calling `xlsx.close`, so reading it afterwards isn't possible. There are workarounds (see below), but it would be nice if this wasn't needed.

It seems like it's the zip-library that closes the input_io (the argument to Workbook.new), so it may be outside the scope of this lib.

Let me know what you think!

```ruby
# does not work
def generate_xlsx(original_file)
  xlsx = Xlsxtream::Workbook.new(original_file)

  xlsx.write_worksheet 'Sheet1' do |sheet|
    sheet << ['cat', 'dog', Time.now]
  end

  xlsx.close

  original_file << path.read
  original_file
end

def other_method
  output = generate_xlsx(my_file)
  upload_to_s3(output.read) # raises `IOError: closed stream`
end

# works
def generate_xlsx(original_file)
  path = Rails.root.join('tmp', "xlsx-ephemeral-#{Time.now.to_i}-#{Random.rand}")

  intermediary_file = File.new(path, 'wb')

  xlsx = Xlsxtream::Workbook.new(intermediary_file)

  xlsx.write_worksheet 'Sheet1' do |sheet|
    sheet << ['cat', 'dog', Time.now]
  end

  xlsx.close

  original_file << path.read
  original_file
ensure
  FileUtils.rm path
end

def other_method
  output = generate_xlsx(my_file)
  upload_to_s3(output.read)
end
```

*I know there is an io_wrapper where one could intercept and ignore the closing, but it would be nice if there was an easier way.*

**Comment by felixbuenemann** [comment]
Yes this is caused by rubyzip, but you could use the ˋTempfile#openˋ or ˋIO#reopenˋ instance methods to reopen the file.

The best solution would be to switch to anothe zip library thats more flexible, but I haven't had the time yet to look into it.

**Comment by sandstrom** [comment]
I checked and there doesn't seem to be another viable alternative. I guess the fix would require a change to RubyZip, such that it could be instructed to not close the file (which in turn could be made configurable in Xlsxtream).

**Comment by felixbuenemann** [comment]
@sandstrom Have you tried the suggested workaround of re-opening the file?

If that doesn't work one way to solve it would be to implement a `#flush` (or `#finish`) method on the `Workbook`, which would write all data and empty the compression buffers without closing the underlying file descriptor.

The problem with that approach is that it does not map well to the other Xlsxtream IO providers like `Xlsxtream::IO::Directory` which writes the files normally contained in the ZIP to a directory tree.

**Comment by felixbuenemann** [comment]
I have another idea how this could be handled, that makes more sense:

If `Xlsxtream::IO::Rubyzip` gets initialized with a path, it will take care of closing the IO object that it created, but it if gets called with an existing IO object, it will not close it.

**Comment by sandstrom** [comment]
@felixbuenemann Yes, that sounds like an excellent idea! 💯 

That way one could feed it with a `TempFile` or some other `IO` object.

**Comment by felixbuenemann** [comment]
> If Xlsxtream::IO::Rubyzip gets initialized with a path, it will take care of closing the IO object that it created, but it if gets called with an existing IO object, it will not close it.

I have just released xlsxtream 1.0.0, which implements this change.

**Comment by sandstrom** [comment]
Awesome! 🏅
